File: 每日一题/剑指offer再刷/剑指 Offer 26. 树的子结构.py
# https://leetcode-cn.com/problems/shu-de-zi-jie-gou-lcof/
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        def recur(A: TreeNode, B: TreeNode) -> bool:
            if not B:
                return True
            if not A:
                return False
            if A.val != B.val:
                return False
            return recur(A.left, B.left) and recur(A.right, B.right)

        if (not A) or (not B):
            return False
        return recur(A, B) or self.isSubStructure(A.left, B) or self.isSubStructure(A.right, B)

File: ts-version/剑指 Offer 26. 树的子结构.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
  class TreeNode {
      val: number
      left: TreeNode | null
      right: TreeNode | null
      constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
      }
  }
function isSubStructure(A: TreeNode | null, B: TreeNode | null): boolean {
    function recur(nodeA: TreeNode | null, nodeB: TreeNode | null): boolean {
        // 判断A是否包含B
        // 1. B空，视为包含
        // 2. A空 || A.VAL != B.VAL 视为不包含
        if (!nodeB) return true
        if (!nodeA || nodeA.val !== nodeB.val) return false
        return recur(nodeA.left, nodeB.left) && recur(nodeA.right, nodeB.right)
    }
    
    if (!A || !B) return false

    return recur(A, B) || isSubStructure(A.left, B) || isSubStructure(A.right, B)

};